docs/architecture/adr-038-state-listening.md

```diff
@@ -269,31 +367,45 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 
 	...
 
-	// call the hooks with the BeginBlock messages
+	// call the streaming service hook with the BeginBlock messages
 	wg := new(sync.WaitGroup)
-	for _, streamingListener := range app.abciListeners {
-		streamingListener := streamingListener // https://go.dev/doc/faq#closures_and_goroutines
-		if streamingListener.HaltAppOnDeliveryError() {
-			// increment the wait group counter
+	if app.abciListener != nil {
+		if app.stopNodeOnStreamingErr {
 			wg.Add(1)
 			go func() {
-				// decrement the counter when the go routine completes
 				defer wg.Done()
-				if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
+				reqBz, err := req.Marshal()
+				if err != nil {
+					panic(err)
+				}
+				resBz, err := res.Marshal()
+				if err != nil {
+					panic(err)
+				}
+				blockHeight := app.deliverState.ctx.BlockHeight()
+				if err := app.abciListener.ListenBeginBlock(blockHeight, reqBz, resBz); err != nil {
 					app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
-					app.halt()
+					panic(err)
```

Comment:
Why this change?

Reply:
Based on your comment here. https://github.com/cosmos/cosmos-sdk/pull/11691#discussion_r898490884. Should we go for `os.Exit(1)` instead?

Reply:
Strictly speaking, for programs like Cosmos nodes, nothing should ever call panic or os.Exit or app.halt except func main. Practically speaking, you can't rely on panic to actually terminate the process: it's frequently used as an ersatz error reporting mechanism, which means there are lots of calls to recover that you can't reliably know about, which means your panic is very likely to be intercepted and squelched. If you want the process to terminate you better call os.Exit.